### ISSUE TITLE
Link expander broken

### ISSUE THREAD

**Issue by selfagency** [selfagency/bs-detector#33]
Need to figure out a way to queue requests so it doesn't take down the server

**Comment by saimoomsafayet** [comment]
is the link expander still broken? as i run the latest version of the extension, expandLinks() method seems never get called.

**Comment by selfagency** [comment]
@lewisje did you fix the link expander as well?

**Comment by Bastlynn** [comment]
I think I ended up fixing the link expander in my firefox extension - I didn't realize it was broken in the other so did what I did to get it working in mine. There wasn't a call to expandLinks() that was active, it was commented out in the copy I was looking at last night. We could easily port the fix from the FF extension to the chrome one and see how that works for us?

**Comment by lewisje** [comment]
I tried fixing it, but I kept getting the message that the link couldn't be expanded.

**Comment by Bastlynn** [comment]
There is a rate limit on how often you can expand new links. So maybe that's what you were hitting? https://unshorten.me/api

**Comment by selfagency** [comment]
Yeah, I can't seem to find any API for link expansion that hasn't gone offline or isn't severely rate limited. That's why I built my own (https://github.com/selfagency/bs-detector/blob/master/linkexpander/x.php) but it took down my server within 10 minutes of rolling it out to plugin users. Open to any other suggestions, especially about client-side solutions for that don't require PHP or any kind of server-side assist.

**Comment by Bastlynn** [comment]
I'll do some thinking about that tonight - I would hope that there's a way to do it that isn't just giving the site a free click through.

**Comment by lewisje** [comment]
If it's configured like unshorten.me, it's just one free click-through per shortened link.

**Comment by selfagency** [comment]
Does anyone have any ideas for how to do this browser-side?

**Comment by thinkcontext** [comment]
The problem w/ doing it browser side by visiting links in the background is it would be easy for the user to be tracked.

**Comment by thinkcontext** [comment]
Thinking about this a little more, having the client expand links allows both the shortener service (bitly, etc) and the destination site to track you.  Some shortener services will tell you the destination (with bitly just add a + on the end) so that would somewhat lessen the problem.

**Comment by selfagency** [comment]
what if we make it optional on the settings page and explicitly state the privacy implications of switching it on

**Comment by selfagency** [comment]
another thought— could we use the [fetch api](https://developers.google.com/web/updates/2015/03/introduction-to-fetch) to request specific headers without loading a page fully so at best the second party logs an IP in their server logs but it doesn't turn up on google or any other page analytics?

**Comment by thinkcontext** [comment]
Does fetch let you control whether cookies get set?

I'd also be concerned about the possibility that the link itself contains a tracking token.  Say a user with the extension gets a shortened link with an individually identifiable token in an email and then views it in the browser.  There are probably other scenarios that could happen as well.

**Comment by selfagency** [comment]
as far as i understand, whether xmlhttprequest or fetch, the page comes back as text. since it's not loaded in the browser as an HTML document, none of the javascript on the page actually fires. therefore a cookie can't be set, a tracking link can't send a ping home, etc.

**Comment by lewisje** [comment]
I believe that the server can still set cookies: Cookies are an HTTP thing, not a JS thing; it's just that JS in the browser has an API for interacting with cookies (except for "HTTP-Only" cookies).

**Comment by selfagency** [comment]
tc and i discussed this in the gitter last night and we believe we found a way around it
